tests/fixtures/liteclienthsmtest.cc

```diff
@@ -0,0 +1,672 @@
+
+    boost::format doCsr("OPENSSL_CONF=%s openssl req -new -engine pkcs11 -keyform engine -key %s");
+    cmd = boost::str(doCsr % cnf_ % key);
+    if (Utils::shell(cmd, &out, true) != EXIT_SUCCESS) {
+      LOG_INFO << "Error: " << out;
+      throw std::runtime_error(cmd.c_str());
+    }
+
+    /* write CSR to disk */
+    csr = hsm_->path_ + csr;
+    Utils::writeFile(csr, out, true);
+  }
+
+  void createCrt(std::string csr, std::string& crt) {
+    crt = hsm_->path_ + crt;
+    rootCa_.signCsr(csr, crt, "");
+  }
+
+  void importCrt(std::string& crt, std::string id) {
+    boost::format crtToDer("OPENSSL_CONF=%s openssl x509 -inform pem -in %s -out %s/tmp.der");
+    cmd = boost::str(crtToDer % cnf_ % crt % hsm_->path_);
+    if (Utils::shell(cmd, &out, true) != EXIT_SUCCESS) {
+      LOG_INFO << "Error: " << out;
+      throw std::runtime_error(cmd.c_str());
+    }
+
+    boost::format writeCrtToHsm("pkcs11-tool --module %s -w %s/tmp.der -y cert --id %s --pin %s");
+    cmd = boost::str(writeCrtToHsm % hsm_->module_ % hsm_->path_ % id % hsm_->pin_);
+    if (Utils::shell(cmd, &out, true) != EXIT_SUCCESS) {
+      LOG_INFO << "Error: " << out;
+      throw std::runtime_error(cmd.c_str());
+    }
+  }
+
+  void listInfo() {
+    boost::format listMechanisms("pkcs11-tool --module %s --list-mechanisms");
+    cmd = boost::str(listMechanisms % hsm_->module_);
+    if (Utils::shell(cmd, &out, true) != EXIT_SUCCESS) {
+      LOG_INFO << "Error: " << out;
+      throw std::runtime_error(cmd.c_str());
+    }
+    // very verbose: enable if debug needed
+    // LOG_INFO << out;
+    boost::format listObjects("pkcs11-tool --module %s --list-objects");
+    cmd = boost::str(listObjects % hsm_->module_);
+    if (Utils::shell(cmd, &out, true) != EXIT_SUCCESS) {
+      throw std::runtime_error(cmd.c_str());
+    }
+    // very verbose: enable if debug needed
+    // LOG_INFO << out;
+  }
+
+ private:
+  SoftHsm* hsm_;
+  RootCaPKI& rootCa_;
+  std::string cnf_;
+  /* buffers */
+  std::string cmd;
+  std::string out;
+};
+
+/**
+*  PKI subscriber (LiteClient) registering with the HSM via the deviceHSM
+*
+*/
+class SubscriberPKI {
+ public:
+  SubscriberPKI(DeviceHsm deviceHsm, std::string keyId, std::string certId, std::string keyLabel, std::string csr,
+                std::string crt)
+      : keyId_(std::move(keyId)),
+        certId_(std::move(certId)),
+        keyLabel_(std::move(keyLabel)),
+        csr_{std::move(csr)},
+        crt_{std::move(crt)}
+  {
+    deviceHsm.createKey(keyId_, keyLabel_);
+    deviceHsm.createCsr(keyLabel_, csr_);
+    deviceHsm.createCrt(csr_, crt_);
+    deviceHsm.importCrt(crt_, certId_);
+    /* enable for debug */
+    deviceHsm.listInfo();
+  }
+
+ public:
+  std::string keyId_;
+  std::string certId_;
+
+ private:
+  std::string keyLabel_;
+  std::string csr_;
+  std::string crt_;
+};
+
+class ClientHSMTest : public ::testing::Test {
```

Comment:
I wonder if it's possible to reuse the liteclient fixture https://github.com/foundriesio/aktualizr-lite/blob/7cf3104293638743542cd6e95d18ab8724f63399/tests/fixtures/liteclienttest.cc#L179, in the way it's reused in https://github.com/foundriesio/aktualizr-lite/blob/7cf3104293638743542cd6e95d18ab8724f63399/tests/aklite_test.cc#L19. It might require making the following method virtual ` std::shared_ptr<LiteClient> createLiteClient(const std::shared_ptr<AppEngine>& app_engine,
                                               InitialVersion initial_version = InitialVersion::kOn,
                                               boost::optional<std::vector<std::string>> apps = boost::none,
                                               const std::string& compose_apps_root = "") ` 
                                               
The idea is that `ClientHSMTest` implements just HSM&KPI stuff. Then `LiteClientHSMTest` will do multiple inheretence https://github.com/foundriesio/aktualizr-lite/pull/82/files#diff-520be724fef6b0a0102c23bea7df3d283005bb7d5e81f49765e541c9f2c5d187R55, `class LiteClientHSMTest: public fixtures::ClientTest, public fixtures::HSMTest`

Reply:
sure I can do that. if we are not pressed for time I can rework things quite a bit now that I have a better picture of the end goal.  I could represent the rootCA as a persistent entity along with the server...anyway, ok., let me rework it all a bit

Reply:
I think each mock should also bit on its on file with some -minimal- comments as to what it does. will do that as well

Reply:
@mike-sul have a look

Reply:
@ldts It looks fine at first glance, will check it out thoroughly tomorrow morning.